libutils/threaded_queue.c

```diff
@@ -0,0 +1,461 @@
+                                                   ItemDestroy);
+
+    // enables errorchecking for deadlocks
+    pthread_mutexattr_t attr;
+    pthread_mutexattr_init(&attr);
+    int ret = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to use error-checking mutexes for queue, "
+            "falling back to normal ones (pthread_mutexattr_settype: %s)",
+            GetErrorStrFromCode(ret));
+        pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_NORMAL);
+    }
+
+    queue->lock = malloc(sizeof(pthread_mutex_t));
+    ret = pthread_mutex_init(queue->lock, &attr);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to initialize mutex (pthread_mutex_init: %s)",
+            GetErrorStrFromCode(ret));
+        pthread_mutexattr_destroy(&attr);
+        free(queue->lock);
+        free(queue);
+        return NULL;
+    }
+
+    pthread_mutexattr_destroy(&attr);
+
+    queue->pcond = malloc(sizeof(pthread_cond_t));
+    ret = pthread_cond_init(queue->pcond, NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to initialize thread condition (pthread_cond_init: %s)",
+            GetErrorStrFromCode(ret));
+        free(queue->lock);
+        free(queue->pcond);
+        free(queue);
+        return NULL;
+    }
+
+    return queue;
+}
+
+ThreadedQueue *ThreadedQueueNewNonLock(size_t initial_capacity,
+                                       void (ItemDestroy) (void *item))
+{
+    ThreadedQueue *queue = xmalloc(sizeof(ThreadedQueue));
+
+    if (initial_capacity == 0)
+    {
+        initial_capacity = DEFAULT_CAPACITY;
+    }
+
+    queue->lock = NULL;
+    queue->pcond = NULL;
+    queue->capacity = initial_capacity;
+    queue->head = 0;
+    queue->tail = 0;
+    queue->size = 0;
+    queue->data = xmalloc(sizeof(void *) * initial_capacity);
+    queue->ItemDestroy = ItemDestroy;
+
+    return queue;
+}
+
+void ThreadedQueueDestroy(ThreadedQueue *queue)
+{
+    if (queue != NULL)
+    {
+        _ThreadedQueueLock(queue->lock);
+        _DestroyRange(queue, queue->head, queue->tail);
+        _ThreadedQueueUnlock(queue->lock);
+
+        ThreadedQueueSoftDestroy(queue);
+    }
+}
+
+void ThreadedQueueSoftDestroy(ThreadedQueue *queue)
+{
+    if (queue != NULL)
+    {
+        if (queue->lock != NULL)
+        {
+            pthread_mutex_destroy(queue->lock);
+            free(queue->lock);
+        }
+
+        if (queue->pcond != NULL)
+        {
+            pthread_cond_destroy(queue->pcond);
+            free(queue->pcond);
+        }
+
+        free(queue->data);
+        free(queue);
+    }
+}
+
+void *ThreadedQueuePop(ThreadedQueue *queue)
+{
+    assert(queue != NULL);
+
+    _ThreadedQueueLock(queue->lock);
+    void *item = _PopNonLock(queue);
+    _ThreadedQueueUnlock(queue->lock);
+
+    return item;
+}
+
+void *ThreadedQueuePopBlock(ThreadedQueue *queue)
```

Comment:
It would also be great to have `ThreadedQueuePopN(queue, n, &n_popped)`, which would pop at most `n` elements at once (within a single lock-unlock "session"). If the items are easy to process, the locking overhead can be quite big if just one item is popped at a time.

Reply:
>>> I think these two functions should be merged into a single one with the `timeout` argument determining what should happen -- return immediately (`0`), wait for `timeout` seconds or wait indefinitely (`-1`). And such a function needs to return `true/false` to indicate if it got a value or timed out.

Won't this inhibit how obvious it is to use? I would rather have two functions, where one returns immediately regardless, and the other takes in enums like `THREAD_BLOCK_INDEFINITELY` (-1), `THREAD_BLOCK_NONE` (0), and some other default value as a parameter.

Maybe I'm focusing on the wrong things?

Reply:
I think we are all capable of reading the function's documentation if it's there. So I don't think it's a big deal if we only have one function. But if you want to add really straightforward functions like `PopWait()` and `TryPop()`, I think they should be based on a generic one that has this **properly documented** versatile `int timeout` parameter.